DigitalLearningSolutions.Data/Services/TrackerService.cs

```diff
@@ -65,6 +72,31 @@ public string ProcessQuery(TrackerEndpointQueryParams query)
                         );
                     }
 
+                    if (action == TrackerEndpointAction.StoreAspProgressV2)
+                    {
+                        if (query.ProgressId == null || query.Version == null || query.TutorialId == null ||
+                            query.CandidateId == null || query.CustomisationId == null)
+                        {
+                            return TrackerEndpointResponse.StoreAspProgressV2Exception;
+                        }
+
+                        if (query.TutorialTime == null || query.TutorialStatus == null)
+                        {
+                            return TrackerEndpointResponse.NullTutorialStatusOrTime;
+                        }

```

Comment:
This validation would make more sense inside the tracker action itself. This method is about selecting the correct tracker action, so it shouldn't also be responsible for doing this validation (single responsibility principle).

Reply:
Moved the checked and corresponding unit testing to the TrackerActionService